scripts/o3de/tests/test_enable_gem.py

```diff
@@ -164,3 +190,152 @@ def add_gem_dependency(enable_gem_cmake_file: pathlib.Path, gem_name: str):
+            return project_data
+
+        def get_gem_json_data(gem_path: pathlib.Path, project_path: pathlib.Path):
+            gem_data = self.enable_gem.gem_data.copy()
+            if gem_path.name in gem_names_and_versions:
+                # gem dependencies data
+                gem_data['gem_name'] = gem_path.name 
+                gem_data['gem_version'] = gem_names_and_versions[gem_path.name] 
+            else:
+                # test gem data
+                gem_data['gem_version'] = gem_version
+                gem_data['dependencies'] = gem_dependencies
+                if compatible_engines:
+                    gem_data['compatible_engines'] = compatible_engines
+                if engine_api_dependencies:
+                    gem_data['engine_api_dependencies'] = engine_api_dependencies
+            return gem_data
+
+        def get_project_gems(project_path: pathlib.Path):
+            return [pathlib.Path(gem_path).resolve()] if gem_registered_with_project else []
+
+        def get_engine_gems():
+            gem_paths = list(map(lambda path:pathlib.Path(path).resolve(), gem_names_and_versions.keys()))
+            if gem_registered_with_engine:
+                gem_paths.append(pathlib.Path(gem_path).resolve())
+            return gem_paths
+
+        def add_gem_dependency(enable_gem_cmake_file: pathlib.Path, gem_name: str):
+            return 0
+
+        with patch('pathlib.Path.is_dir', return_value=True) as pathlib_is_dir_patch,\
+                patch('pathlib.Path.is_file', return_value=True) as pathlib_is_file_patch, \
+                patch('o3de.manifest.load_o3de_manifest', side_effect=load_o3de_manifest) as load_o3de_manifest_patch, \
+                patch('o3de.manifest.save_o3de_manifest', side_effect=save_o3de_manifest) as save_o3de_manifest_patch,\
+                patch('o3de.manifest.get_registered', side_effect=get_registered_path) as get_registered_patch,\
+                patch('o3de.manifest.get_engine_json_data', side_effect=get_engine_json_data) as get_engine_json_data_patch,\
+                patch('o3de.manifest.get_gem_json_data', side_effect=get_gem_json_data) as get_gem_json_data_patch,\
+                patch('o3de.manifest.get_project_json_data', side_effect=get_project_json_data) as get_gem_json_data_patch,\
+                patch('o3de.manifest.get_project_gems', side_effect=get_project_gems) as get_project_gems_patch,\
+                patch('o3de.manifest.get_engine_gems', side_effect=get_engine_gems) as get_engine_gems_patch,\
+                patch('o3de.cmake.add_gem_dependency', side_effect=add_gem_dependency) as add_gem_dependency_patch,\
+                patch('o3de.validation.valid_o3de_gem_json', return_value=True) as valid_gem_json_patch:
+
+            self.enable_gem.project_data.pop('gem_names', None)
+            result = enable_gem.enable_gem_in_project(gem_path=gem_path, project_path=project_path, force=force, check=check)
```

Comment:
Does this work when the optional gem flag is supplied?

Reply:
This brings up a good question on what the default behavior should be if you try to add an optional gem that is incompatible.  For now one option we could take would be to have the operation fail if you try to add an optional incompatible gem so you see the error message - you can always use `force` to enable it despite the incompatibility.  Thoughts @lumberyard-employee-dm ?

Reply:
That sounds fine with me for checking incompatibility.
I see the common use case as the user adding the optional gem would have the gem available on their machine, but inspecting others users to have the gem available.
Therefore checking for incompatibility when the `--is-optional-gem` flag is supplied makes sense.

Now on to clarify what I was actually asking.
I was wondering if the "gem_names" entry was actually updated correctly with the JSON object that specifies optional when checking for incompatiblity.
i.e
```json
"gem_names" : [
    "atom=1.2.3",
    {
        "name": "openxr=2.3.4",
        "optional": true
    }
]
```

Reply:
I added unit tests to verify you get a dictionary object with `name` and `optional` when adding an optional gem